File: 3777.minimum-deletions-to-make-alternating-substring.go
#
# @lc app=leetcode id=3777 lang=golang
#
# [3777] Minimum Deletions to Make Alternating Substring
#

# @lc code=start
func minDeletions(s string, queries [][]int) []int {
	n := len(s)
	m := n - 1
	if m < 0 {
		var ans []int
		for _, q := range queries {
			if q[0] == 2 {
				ans = append(ans, 0)
			}
		}
		return ans
	}
	tree := make([]int, m+2)
	update := func(idx int, delta int) {
		idx++ // 1-based
		for idx <= m {
			tree[idx] += delta
			idx += idx & -idx
		}
	}
	prefix := func(idx int) int {
		idx++ // 1-based
		res := 0
		for idx > 0 {
			res += tree[idx]
			idx -= idx & -idx
		}
		return res
	}
	rangeSum := func(left, right int) int {
		if left > right {
			return 0
		}
		res := prefix(right)
		if left > 0 {
			res -= prefix(left - 1)
		}
		return res
	}
	// Init BIT
	for i := 0; i < m; i++ {
		d := 0
		if s[i] != s[i+1] {
			d = 1
		}
		update(i, d)
	}
	var ans []int
	for _, q := range queries {
		tp := q[0]
		if tp == 1 {
			j := q[1]
			if j > 0 {
				curr := rangeSum(j-1, j-1)
				delta := 1 - 2 * curr
				update(j-1, delta)
			}
			if j < n-1 {
				curr := rangeSum(j, j)
				delta := 1 - 2 * curr
				update(j, delta)
			}
		} else {
			l, r := q[1], q[2]
			length := r - l + 1
			if length == 1 {
				ans = append(ans, 0)
				continue
			}
			sumd := rangeSum(l, r-1)
			dels := length - 1 - sumd
			ans = append(ans, dels)
		}
	}
	return ans
}
# @lc code=end

File: 3777.minimum-deletions-to-make-alternating-substring.java
#
# @lc app=leetcode id=3777 lang=java
#
# [3777] Minimum Deletions to Make Alternating Substring
#

# @lc code=start
class Solution {
    private static class FenwickTree {
        private int[] tree;
        private int n;

        public FenwickTree(int size) {
            this.n = size;
            this.tree = new int[size + 1];
        }

        public void update(int index, int delta) {
            while (index <= this.n) {
                this.tree[index] += delta;
                index += index & -index;
            }
        }

        public int query(int index) {
            int sum = 0;
            while (index > 0) {
                sum += this.tree[index];
                index -= index & -index;
            }
            return sum;
        }

        public int query(int left, int right) {
            return this.query(right) - this.query(left - 1);
        }
    }

    public int[] minDeletions(String s, int[][] queries) {
        int n = s.length();
        char[] ch = s.toCharArray();
        FenwickTree ft = new FenwickTree(n - 1);
        for (int i = 0; i < n - 1; ++i) {
            int d = (ch[i] != ch[i + 1]) ? 1 : 0;
            ft.update(i + 1, d);
        }
        int[] answer = new int[queries.length];
        int ansIdx = 0;
        for (int[] q : queries) {
            if (q[0] == 1) {
                int j = q[1];
                int old_left = (j > 0) ? ((ch[j - 1] != ch[j]) ? 1 : 0) : 0;
                int old_right = (j < n - 1) ? ((ch[j] != ch[j + 1]) ? 1 : 0) : 0;
                ch[j] = (ch[j] == 'A') ? 'B' : 'A';
                int new_left = (j > 0) ? ((ch[j - 1] != ch[j]) ? 1 : 0) : 0;
                int new_right = (j < n - 1) ? ((ch[j] != ch[j + 1]) ? 1 : 0) : 0;
                if (j > 0) {
                    ft.update(j, new_left - old_left);
                }
                if (j < n - 1) {
                    ft.update(j + 1, new_right - old_right);
                }
            } else {
                int l = q[1];
                int r = q[2];
                int sum_diff = (l < r) ? ft.query(l + 1, r) : 0;
                int del = (r - l) - sum_diff;
                answer[ansIdx++] = del;
            }
        }
        return java.util.Arrays.copyOf(answer, ansIdx);
    }
}
# @lc code=end